### PR TITLE
minor: add custom tokens for Disjunctive Normal Form types parentheses

### PR DESCRIPTION
**Pull request by kubawerlos**

### CODE REVIEW

File: tests/Tokenizer/TokensTest.php [hunk 1]
```diff
@@ -741,6 +741,43 @@ public static function provideFindBlockEnd80Cases(): array
...
+     * @dataProvider provideFindBlockEnd82Cases
+     *
+     * @param Tokens::BLOCK_TYPE_* $type
+     */
+    public function testFindBlockEnd82(int $expectedIndex, string $source, int $type, int $searchIndex): void
+    {
+        static::assertFindBlockEnd($expectedIndex, $source, $type, $searchIndex);
+    }
+
+    public static function provideFindBlockEnd82Cases(): iterable
```
Comment (Wirone) [inline tests/Tokenizer/TokensTest.php:756]: Shouldn't there be test cases for DNF types in arrow functions and/or class' properties?

Similar in `DisjunctiveNormalFormTypeParenthesisTransformerTest`.

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+     * @param array<int, int> $expectedTokens
+     *
+     * @dataProvider provideProcessCases
+     */
+    public function testProcess(string $source, array $expectedTokens): void
+    {
+        $this->doTest($source, $expectedTokens, [CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN, CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE]);
+    }
+
+    public static function provideProcessCases(): iterable
```
Comment (Wirone) [inline tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:39]: Maybe we should include invokable class in test case? There are a lot of `(`+`)` pairs that shouldn't be matched as DNFs - [example](https://3v4l.org/aTb7e).

```php
interface Foo {}
interface Bar {}

class A {
    public function __invoke((Foo&Bar)|string $a): string {
        return 'A';
    }
}

return (new A())(new class implements Foo, Bar {});
```

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                63 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+                81 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN, // $uu
+                85 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+                140 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN, // $f
+                144 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+            ],
+        ];
+
+        yield 'one line function signatures with comments' => [
+            '<?php function Foo1 ((A&B)|C $x1, \C|(A&B) $x2,/**/(A&B)|I $xxx): void {}',
```
Comment (Wirone) [inline tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:71]: Maybe we should put DNF-like inside the comment and ensure it's not matched as DNF?

```suggestion
            '<?php function Foo1 ((A&B)|C $x1, \C|(A&B) $x2,/* No (X&Y) ! */(A&B)|I $xxx): void {}',
```

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                11 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN,
+                15 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+            ],
+        ];
+
+        yield 'DNF property, (E&F)|G' => [
+            '<?php
+class Dnf
+{
+    public A|(C&D) $a;
```
Comment (Wirone) [inline tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:175]: Shouldn't it be like this?
```suggestion
    public (E&F)|G $a;
```

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+                47 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN,
+                51 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+                58 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN,
+                62 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+                73 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_OPEN,
+                77 => CT::T_DISJUNCTIVE_NORMAL_FORM_TYPE_PARENTHESIS_CLOSE,
+            ],
+        ];
+
+        yield 'constructor promotion' => [
```
Comment (Wirone) [inline tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:234]:
```suggestion
        yield 'constructor property promotion' => [
```

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+        yield 'bigger set of multiple DNF properties' => [
+            '<?php
+class Dnf
+{
+    public A|(C&D) $a;
+    protected (C&D)|B $b;
+    private (C&D)|(E&F)|(G&H) $c;
+    static (C&D)|Z $d;
+    public /* */ (C&D)|X $e;
```
Comment (Wirone) [inline tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:206]: Maybe DNF-like inside the comment?

**Review by Wirone — CHANGES_REQUESTED** [review]

File: tests/Tokenizer/TokensTest.php (hunk 1 above)
Comment (kubawerlos) [reply tests/Tokenizer/TokensTest.php:756]: If we test it in `DisjunctiveNormalFormTypeParenthesisTransformerTest` will adding the checks here bring any value?

Comment (Wirone) [reply tests/Tokenizer/TokensTest.php:756]: I don't know 🤷‍♂️😅. You know better probably, I think at least in one place would be enough.

File: tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php (hunk 6 above)
Comment (kubawerlos) [reply tests/Tokenizer/Transformer/DisjunctiveNormalFormTypeParenthesisTransformerTest.php:206]: Let's have it once in comment and once in PHPDoc.

**Review by Wirone — APPROVED** [review]

**Comment by keradus** [comment]
kudos @kubawerlos for PR, @SpacePossum for initial work on it, and @Wirone for great review!

**Pull request merged**